wagtail/wagtailusers/views/groups.py

```diff
@@ -26,133 +22,152 @@ def get_permission_panel_classes():
     return _permission_panel_classes
 
 
-@any_permission_required('auth.add_group', 'auth.change_group', 'auth.delete_group')
-@vary_on_headers('X-Requested-With')
-def index(request):
-    q = None
-    is_searching = False
-
-    if 'q' in request.GET:
-        form = SearchForm(request.GET, placeholder=_("Search groups"))
-        if form.is_valid():
-            q = form.cleaned_data['q']
-
-            is_searching = True
-            groups = Group.objects.filter(name__icontains=q)
-    else:
-        form = SearchForm(placeholder=_("Search groups"))
-
-    if not is_searching:
-        groups = Group.objects
-
-    groups = groups.order_by('name')
-
-    if 'ordering' in request.GET:
-        ordering = request.GET['ordering']
-
-        if ordering in ['name', 'username']:
-            if ordering != 'name':
-                groups = groups.order_by(ordering)
-    else:
-        ordering = 'name'
-
-    paginator, groups = paginate(request, groups)
-
-    if request.is_ajax():
-        return render(request, "wagtailusers/groups/results.html", {
-            'groups': groups,
-            'is_searching': is_searching,
-            'query_string': q,
-            'ordering': ordering,
-        })
-    else:
-        return render(request, "wagtailusers/groups/index.html", {
-            'search_form': form,
-            'groups': groups,
-            'is_searching': is_searching,
-            'ordering': ordering,
-            'query_string': q,
-        })
-
-
-@permission_required('auth.add_group')
-def create(request):
-    group = Group()
-    if request.method == 'POST':
-        form = GroupForm(request.POST, instance=group)
-        permission_panels = [
-            cls(request.POST, instance=group)
+class PermissionPanelFormsMixin(object):
+    def get_permission_panel_form_kwargs(self, cls):
+        kwargs = {}
+
+        if self.request.method in ('POST', 'PUT'):
+            kwargs.update({
+                'data': self.request.POST,
+                'files': self.request.FILES,
+            })
+
+        if hasattr(self, 'object'):
+            kwargs.update({'instance': self.object})
+
+        return kwargs
+
+    def get_permission_panel_forms(self):
+        return [
+            cls(**self.get_permission_panel_form_kwargs(cls))
             for cls in get_permission_panel_classes()
         ]
+
+    def get_context_data(self, **kwargs):
+        if 'permission_panels' not in kwargs:
+            kwargs['permission_panels'] = self.get_permission_panel_forms()
+
+        return super(PermissionPanelFormsMixin, self).get_context_data(**kwargs)
+
+
+class SearchableListMixin(object):
```

Comment:
Would `wagtail.wagtailadmin.views.generic` be a better home for `SearchableListMixin`?

Reply:
I was reluctant to do this because it uses ``__icontains`` and this isn't the best method of searching for all models. I think a generic implementation should use ``wagtailsearch`` if it's configured on the model being searched so left this implementation here for now.